components/brave_wallet/browser/zcash/zcash_blocks_batch_scan_task.cc

```diff
@@ -0,0 +1,256 @@
+  ScheduleWorkOnTask();
+}
+
+void ZCashBlocksBatchScanTask::GetFrontierBlock() {
+  context_->zcash_rpc->GetCompactBlocks(
+      context_->chain_id, frontier_block_height_, frontier_block_height_,
+      base::BindOnce(&ZCashBlocksBatchScanTask::OnGetFrontierBlock,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashBlocksBatchScanTask::OnGetFrontierBlock(
+    base::expected<std::vector<zcash::mojom::CompactBlockPtr>, std::string>
+        result) {
+  if (!result.has_value() || result.value().size() != 1) {
+    error_ = ZCashShieldSyncService::Error{
+        ZCashShieldSyncService::ErrorCode::kFailedToDownloadBlocks,
+        result.error()};
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  frontier_block_ = std::move(result.value()[0]);
+  ScheduleWorkOnTask();
+}
+
+void ZCashBlocksBatchScanTask::DownloadBlocks() {
+  uint32_t start_index =
+      downloaded_blocks_ ? from_ + downloaded_blocks_->size() : from_;
+  uint32_t end_index = std::min(start_index + kBlockDownloadBatchSize - 1, to_);
+  auto expected_size = end_index - start_index + 1;
+
+  context_->zcash_rpc->GetCompactBlocks(
+      context_->chain_id, start_index, end_index,
+      base::BindOnce(&ZCashBlocksBatchScanTask::OnBlocksDownloaded,
+                     weak_ptr_factory_.GetWeakPtr(), expected_size));
+}
+
+void ZCashBlocksBatchScanTask::OnBlocksDownloaded(
+    size_t expected_size,
+    base::expected<std::vector<zcash::mojom::CompactBlockPtr>, std::string>
+        result) {
+  CHECK(frontier_block_);
+  CHECK(frontier_tree_state_);
+  if (!result.has_value()) {
+    error_ = ZCashShieldSyncService::Error{
+        ZCashShieldSyncService::ErrorCode::kFailedToDownloadBlocks,
+        result.error()};
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  if (expected_size != result.value().size()) {
+    error_ = ZCashShieldSyncService::Error{
+        ZCashShieldSyncService::ErrorCode::kFailedToDownloadBlocks,
+        "Expected block count doesn't match actual"};
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  if (!downloaded_blocks_) {
+    downloaded_blocks_ = std::vector<zcash::mojom::CompactBlockPtr>();
+  }
+  base::Extend(downloaded_blocks_.value(), std::move(result.value()));
+  ScheduleWorkOnTask();
+}
+
+void ZCashBlocksBatchScanTask::ScanBlocks() {
+  if (!downloaded_blocks_ || downloaded_blocks_->empty()) {
+    error_ = ZCashShieldSyncService::Error{
+        ZCashShieldSyncService::ErrorCode::kScannerError, "No blocks to scan"};
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  if (!frontier_block_.value() || !frontier_tree_state_.value() ||
+      !frontier_block_.value()->chain_metadata) {
+    error_ = ZCashShieldSyncService::Error{
+        ZCashShieldSyncService::ErrorCode::kScannerError, "Frontier error"};
+    ScheduleWorkOnTask();
+    return;
+  }
+
+  OrchardTreeState tree_state;
+  {
+    auto frontier_bytes = PrefixedHexStringToBytes(
+        base::StrCat({"0x", frontier_tree_state_.value()->orchardTree}));
```

Comment:
It looks like it should be `HexStringToBytes(frontier_tree_state_.value()->orchardTree)`

Reply:
PrefixedHexStringToBytes is a bit more complicated: for ex. it appends 0 if input size is odd. I'd prefer to use it instead.

Reply:
PrefixedHexStringToBytes is for EVM's encoding of integers. It is not supposed to be used unless it is explictly required by format of hex string.
As afar as I can see orchardTree is just hex encoded bytes, so decoding to bytes should be done with HexStringToBytes and fail in case of odd size

Reply:
Seems it should be renamed then

Reply:
Switched to HexStringToBytes